### ISSUE TITLE
Adjusting access to utils for federated execution

### ISSUE THREAD

**Issue by hokeun** [lf-lang/reactor-ts#112]
After fed-gen is done, the generated .lf files from a federated LF program will be the main reactors, which is currently generated as `App` class rather than `FederatedApp` class.
However, inside the network sender body (reaction body), `this.util.sendRTITimedMessage` will be called in the main reactor. This is currently not possible since `App` class doesn't implement sendRTIMessage or sendRTITimedMessage like this.

```
    protected sendRTIMessage<T extends Present>(data: T, destFederateID: number, destPortID: number) {
        throw new Error("Cannot call sendRTIMessage from an App. sendRTIMessage may be called only from a FederatedApp");
    }

    protected sendRTITimedMessage<T extends Present>(data: T, destFederateID: number, destPortID: number) {
        throw new Error("Cannot call sendRTIMessage from an App. sendRTIMessage may be called only from a FederatedApp");
    }
```

Possible solutions:

1. Somehow we distinguish a reactor to be generated as `FederatedApp` and a reactor to be generated as `App`.
2. Allow sendRTITimedMessage to be called inside `App`
3. Other suggestions? (@lhstrh)

**Comment by ByeongGil-Jun** [comment]
@marten Below is what I found about waiting `networkInputControlReaction` by using `async`/`await`. It would be a great pleasure if you read it and tell me your thoughts.

There are two problems. 
First, how can we suspend and control back in the right way? 
Second, how can we scan and process the RTI messages while the `_react` is suspended?

For the first question, we should make all upper functions of `_react()` be an async function and put `await` carefully.
These are four scenarios that `_react()` is called.
1. [`_startExecuting`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/reactor.ts#L2339) → `_react()`
2. [`_startExecuting`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/reactor.ts#L2342) → [`_next()`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/reactor.ts#L2011) → `_react()`
3. [`TagAdvanceGrant`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/federation.ts#L1275) or [`ProvisionalTagAdvanceGrant`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/federation.ts#L1290) → [`_requestImmediateInvocationOfNext()`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/reactor.ts#L2158) →[`setImmediate()`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/reactor.ts#L2161) → `_next()` → `_react()`
4. [`setAlarmOrYield`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/reactor.ts#L2101) → [`alarm.set()`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/reactor.ts#L2146) → `_next` → `_react()`

It would be possible if these complex connections are just obstacles. However, we have to consider [`RTIClient`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/federation.ts#L275), too. Because the main loop of federated execution is this [`loop`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/federation.ts#L641). For example, at the start tag, this [`emit code`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/federation.ts#L674) is executed and this [`on function`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/federation.ts#L1198) is invoked. After all codes in that function are executed, `RTI` can invoke other emit codes to make `FederatedApp` process them. So we should change the function of `RTIClient`, too. 

And the second problem raises here, dealing with `RTIClient`. The `FederateApp` can't process another message from `RTI` before it finishes emitting code in [`handleSocketData`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/federation.ts#L674). For now, this function is called only at [`this code`](https://github.com/lf-lang/reactor-ts/blob/554a86128c35c76335dec50d2c953590fb02d259/src/core/federation.ts#L373). If we turn `_react` to an async function and put await, it should get a promise return value from that `handleSocketData` function. The problem is, when we call `handleSocketData` from `_react`, it might execute `_next` again, and we should prevent that happen. Whole functions, emitting message processing codes, `_next()` and `_react()`, `_requestImmediateInvocationOfNext()` are intricately connected.
I think that we should change the whole logic of the federation.ts to use the `async`/`await`.

What is the better way to address this waiting problem, keep my previous solution, or start the discussion to apply `async`/`await`?

**Comment by lhstrh** [comment]
@byeong-gil: I think this may be part-way addressed? I believe we do use the `FederatedApp` class, but there are still methods in the `App` class that belong in the `FederatedApp` class.

**Comment by byeong-gil** [comment]
> @byeong-gil: I think this may be part-way addressed? I believe we do use the `FederatedApp` class, but there are still methods in the `App` class that belong in the `FederatedApp` class.

Yes, I think so. We do use the `FederatedApp` class and the access of util functions like `sendRTITImedMessage` is handled properly.

**Comment by lhstrh** [comment]
But aren't they defined in App (vs. FederatedApp)?

**Comment by byeong-gil** [comment]
Yes, they are. Should we define those things in `FederatedApp`? 

Actually, those functions are defined in [`util`](https://github.com/lf-lang/reactor-ts/blob/bb9b0ea70d0148571e6fdae99c010052e4df9d42/src/core/reactor.ts#L1855) as below and are overridden [here](https://github.com/lf-lang/reactor-ts/blob/bb9b0ea70d0148571e6fdae99c010052e4df9d42/src/core/federation.ts#L1744). I couldn't imagine how we can get rid of them from the `util` class.
https://github.com/lf-lang/reactor-ts/blob/bb9b0ea70d0148571e6fdae99c010052e4df9d42/src/core/reactor.ts#L1916-L1923

**Comment by lhstrh** [comment]
I don't think that `UtilityFunctions` has to be defined anonymously. We can define it in some other file and have a `FederateUtilityFunctions extends UtilityFunctions` that includes all the additional communication methods. We'll just instantiate it in `FederatedApp` (which, in hindsight, I think should actually just be called `Federate`).

**Comment by byeong-gil** [comment]
I got it. I'll make a PR to address it and another PR or issue for renaming (e.g. `FederatedApp` -> `Federate`, `TimedMessage` -> `TaggedMessage`...).
